Qrc20 #621

sergeyboyko0791 opened the pull request:
Basic QRC20 functionality that was implemented:
* electrum call, withdraw, my_balance

Change list:
* UTXO based coins now are enabled with the additional argument:
`"protocol": "UTXO"`
* ETH coin now is enabled with the additional argument:
`"protocol": "ETH"`
* ERC20 based coins now are enabled with the additional argument:
`"protocol": { "platform": "ETH", "contract_address": 0x[H160] }`
* QRC20 based coins now are enabled with the additional argument:
`"protocol": { "platform": "QTUM", "contract_address": H160 }`

artemii235 commented on mm2src/coins/utxo.rs:
```diff
@@ -256,1602 +254,151 @@ pub struct UtxoCoinImpl {  // pImpl idiom.
-#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
-pub struct UtxoFeeDetails {
-    amount: BigDecimal,
+pub trait UtxoArcGetter {
```
Standard AsRef trait looks like good replacement for this: https://doc.rust-lang.org/std/convert/trait.AsRef.html

sergeyboyko0791 replied:
I thought we decide to make a custom trait instead of Deref<UtxoCoinFields>.
What's the best way?

artemii235 replied:
`AsRef` and `Deref` are different things 🙂 While `Deref` is recommended to be used only for smart pointers `AsRef` is `reference-to-reference conversion` which looks fine in this case.

artemii235 commented on mm2src/coins/utxo/qrc20.rs:
```diff
@@ -0,0 +1,690 @@
+    let (unsigned, data) = try_s!(coin.generate_transaction(unspents, outputs, fee_policy, actual_tx_fee, gas_fee).await);
+    let prev_script = Builder::build_p2pkh(&coin.arc().my_address.hash);
+    let signed = try_s!(sign_tx(unsigned, &coin.arc().key_pair, prev_script, coin.arc().signature_version, coin.arc().fork_id));
+    let fee_details = UtxoFeeDetails {
```
UtxoFeeDetails is expected that fee is paid in coin that is withdrawn, but it is not the case for QRC20. Seems like it should be mix of UtxoFeeDetails and EthFeeDetails, e.g.:
```
coin: QTUM, // QRC20 transfer fee is paid in QTUM
miner_fee: standard UTXO miner fee based on transaction size,
gas_price: 40, // satoshis
gas_limit: 250000,
total_gas_fee: 0.1
```

artemii235 commented on mm2src/coins/utxo/qrc20.rs:
```diff
@@ -0,0 +1,690 @@
+    // sometimes: https://etherscan.io/tx/0x4f719da4e138bd8ab929f4110e84d773b57376b37d1c635d26cd263d65da99cb
+    // https://medium.com/chainsecurity/zero-gas-price-transactions-what-they-do-who-creates-them-and-why-they-might-impact-scalability-aeb6487b8bb0
+    // I suspect that there might be an implementation error in contract_encode_number function in
+    // qtum electrum wallet.
```
Looks like QTUM wallets do not allow setting gas price to 0 so we should consider it as not supported. Please remove this TODO.

artemii235 commented on mm2src/coins/utxo.rs:
```diff
@@ -2228,3 +782,104 @@ fn kmd_interest(height: Option<u64>, value: u64, lock_time: u64, current_time: u
+                                              -> Result<UtxoTx, String>
+    where T: UtxoArcGetter + UtxoArcCommonOps {
+    let _utxo_lock = UTXO_LOCK.lock().await;
+    // TODO replace list_unspent_ordered with ordered_mature_unspents
```
TODO?

sergeyboyko0791 replied:
The ordered_mature_unspents method takes MmArc to get path to the database directory, because of this we should add the MmArc to all of this method callers: send_taker_fee, send_maker_payment, send_taker_payment. I think it can be done in the next step - QRC20 swap integration. Or we can do not replace list_unspent_ordered with ordered_mature_unspents.

artemii235 replied:
As we will move coins crate to separate lib some day I would prefer for coins implementation to not directly depend on `MmCtx/MmArc` global state. Can we replace MmArc with something like `cache_directory` field in coin struct? 
It will be set up upon coin initialization using DB dir from MmArc in AtomicDEX-API, but will be also possible to be set to something different. Might be also useful for tests because we won't need to build up entire MmArc for coins specific tests.

artemii235 commented on mm2src/coins/utxo.rs:
```diff
@@ -2228,3 +782,104 @@ fn kmd_interest(height: Option<u64>, value: u64, lock_time: u64, current_time: u
+    let _utxo_lock = UTXO_LOCK.lock().await;
+    // TODO replace list_unspent_ordered with ordered_mature_unspents
+    let unspents = try_s!(coin.arc().rpc_client.list_unspent_ordered(&coin.arc().my_address).map_err(|e| ERRL!("{}", e)).compat().await);
+    // TODO check for QRC20
```
TODO?

sergeyboyko0791 replied:
The generate_transaction method takes `gas_fee:  Optional<u64>`. We can use default gas limit and gas price for the swaps like [here](https://github.com/KomodoPlatform/atomicDEX-API/blob/qrc20/mm2src/coins/utxo/qrc20.rs#L458).

artemii235 commented on mm2src/coins/utxo/qrc20.rs:
```diff
@@ -0,0 +1,690 @@
+
+    fn blockchain_contract_call(&self, contract_addr: &H160Json, data: BytesJson) -> RpcRes<ContractCallResult>;
+}
+
```
Cipi mentioned in issue that QTUM has recently added the docs for QRC20 integration: https://github.com/qtumproject/qtum-electrumx-server/blob/master/docs/qrc20-integration.md, please add the link here.

sergeyboyko0791 replied:
Unfortunately there is not info about blockchain.contract.call.
I've added the link only for the blockchain.token.get_info

artemii235 commented on cmake-build-debug/CMakeFiles/3.16.5/CMakeCCompiler.cmake:
```diff
@@ -0,0 +1,76 @@
+set(CMAKE_C_COMPILER "/usr/bin/cc")
```
Please remove these autogenerated `cmake_build_debug` files from git :slightly_smiling_face:

sergeyboyko0791 replied:
Done

artemii235 commented on mm2src/coins/utxo/qrc20.rs:
```diff
@@ -0,0 +1,698 @@
+async fn qrc20_withdraw(coin: Qrc20Coin, req: WithdrawRequest) -> Result<TransactionDetails, String> {
+    let to_addr = match &coin.utxo_arc.address_format {
+        UtxoAddressFormat::Standard => try_s!(Address::from_str(&req.to)),
+        UtxoAddressFormat::CashAddress { .. } => try_s!(Address::from_cashaddress(
```
QTUM/QRC20 do not support cash addresses for sure, we can simply use `Address::from_str` here and also maybe deny QRC20 coin initialization if it has cash address format in config for some reason.

sergeyboyko0791 replied:
Remove the cashaddress for QRC20

artemii235 commented:
@tonymorony 
Could you please test following on testnet and mainnet:
1. Old coins config is properly converted to new format used by MM2. Docs of coins converter can be seen in `mm2 help`.
2. QRC20 is successfully activated similarly to this test: https://github.com/KomodoPlatform/atomicDEX-API/pull/672/files#diff-5059b1a76e237fd84822c2c8f349455aR2497 (coin config sample and electrum address for testnet are there in code).
3. QRC20 can be successfully withdrawn to another address.
4. `buy/sell/setprice` calls are rejected if 1 of coins is QRC20.

Please note that tx_history doesn't work yet for QRC20.

You can create new QTUM wallet here: https://qtumwallet.org/, please note that for testnet you should first select `testnet` mode in settings.

You can get testnet QTUM from this faucet: http://testnet-faucet.qtum.info/

I can send you some of mine QRC20 token, but you can also deploy your own using web wallet.

tonymorony commented:
Started testing :)  coins file converted just fine

Deployed my token: https://testnet.qtum.info/qrc20/6029bddd12a5a828c89b61dc8c1d607d7e52b182/


made such coins file:

```
[{"coin":"QRC20","required_confirmations":0,"pubtype": 120,"p2shtype": 50,"wiftype": 128,"segwit": true,"txfee": 0,"mm2": 1, "protocol":{"platform":"QTUM","token_type": "QRC20","params": {"contract_address":"0x6029bddd12a5a828c89b61dc8c1d607d7e52b182"}}}]
```

and trying to activate but getting an error:

```
curl --url "http://127.0.0.1:7783" --data "{\"userpass\":\"$userpass\",\"method\":\"electrum\",\"coin\":\"QRC20\",\"servers\":[{\"url\":\"95.217.83.126:10001\"}],\"mm2\":1}"
{"error":"rpc:305] lp_commands:74] lp_coins:661] invalid value: map, expected map with a single key"}
```

probably I've made a mistake somewhere 🤔

sergeyboyko0791 commented:
The protocol field looks like this:

"protocol":{"QRC20":{"platform":"QTUM","contract_address":"0x6029bddd12a5a828c89b61dc8c1d607d7e52b182"}}

You've used an old format, it has been changed recently

tonymorony commented:
> "protocol":{"QRC20":{"platform":"QTUM","contract_address":"0x6029bddd12a5a828c89b61dc8c1d607d7e52b182"}}

thanks @sergeyboyko0791 ! made some progress:

```
curl --url "http://127.0.0.1:7783" --data "{\"userpass\":\"$userpass\",\"method\":\"electrum\",\"coin\":\"QRC20\",\"servers\":[{\"url\":\"95.217.83.126:10001\"}],\"mm2\":1}"
{"error":"rpc:305] lp_commands:75] qrc20:332] JsonRpcError { client_info: \"coin: QRC20\", request: JsonRpcRequest { jsonrpc: \"2.0\", id: \"3\", method: \"blockchain.contract.call\", params: [String(\"6029bddd12a5a828c89b61dc8c1d607d7e52b182\"), String(\"70a08231000000000000000000000000a19f7aa254dced0a17920ab1c2acf698841f67f6\"), String(\"\")] }, error: Response(95.217.83.126:10001, Object({\"code\": Number(2), \"message\": String(\"daemon error: DaemonError({\\'code\\': -5, \\'message\\': \\'Address does not exist\\'})\")})) }"}
```

sergeyboyko0791 commented:
Please check if the MM2 connects to Qtum testnet: {"url":"95.217.83.126:10001"}

tonymorony commented:
>Please check if the MM2 connects to Qtum testnet: {"url":"95.217.83.126:10001"}

yup, I'm using this one

I've noticed that getting: 

```
01 07:01:54, lp_coins:675] ticker = "QRC20", block_count = 625027
```

when actual testnet height is 626257, probably daemon on testnet electrum stuck/hanged and thats why that tx doesn't exists on electrum CC @artemii235

artemii235 commented:
New QTUM version was released 10 days ago, this includes hardfork on testnet on block 625000 (https://github.com/qtumproject/qtum/releases/tag/mainnet-ignition-v0.19.1), updating daemon now.

artemii235 commented:
@tonymorony Could you retry now please? Daemon is updated and synced.

tonymorony commented:
@artemii235 

```
{"result":"success","address":"qYHxziMubYKZVwbKwrsn3YUhW4iL927NAS","balance":"0","unspendable_balance":"0","locked_by_swaps":[[0,[]],[1,[1]]],"coin":"QRC20","required_confirmations":0,"requires_notarization":false}
```

yay, works fine now 💯

tonymorony commented:
```
{"coin":"QRC20","balance":"777","unspendable_balance":"0","locked_by_swaps":{"numer":"0","denom":"1"},"address":"qYHxziMubYKZVwbKwrsn3YUhW4iL927NAS"}
```

```
curl --url "http://127.0.0.1:7783" --data "{\"method\":\"withdraw\",\"coin\":\"QRC20\",\"to\":\"qMHa3L2VG9ZHJYQSZmCpv5PCvtxXfFEyqJ\",\"amount\":\"10\",\"userpass\":\"$userpass\"}"
{"tx_hex":"0100000001747c4c8d3c57e969c9b596726ca8db234828b6f3015bcd922b23ca46354d4926000000006a47304402205b8f7e3f16ec0e3a6752088ad3f0e348fd1ea60128913b5276b688e254624a3f022004763e91d769fb16d7af50865438797750d880fb8a589bdf5dd546ff007b1678012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff02000000000000000062540390d003012844a9059cbb00000000000000000000000028e2ef01b3046a4eaaae78640285a29089bc0b46000000000000000000000000000000000000000000000000000000003b9aca00146029bddd12a5a828c89b61dc8c1d607d7e52b182c255cd341d000000001976a914a19f7aa254dced0a17920ab1c2acf698841f67f688acc962fc5e","tx_hash":"315c3ef52fea59f137bd81581e974710caa89d202e80e32820ab9c7e0fd38896","from":["qYHxziMubYKZVwbKwrsn3YUhW4iL927NAS"],"to":["qMHa3L2VG9ZHJYQSZmCpv5PCvtxXfFEyqJ"],"total_amount":"5","spent_by_me":"5","received_by_me":"4.89999701","my_balance_change":"-0.10000299","block_height":0,"timestamp":1593598665,"fee_details":{"coin":"QRC20","miner_fee":"0.10000299","gas_limit":250000,"gas_price":40,"total_gas_fee":"0.1"},"coin":"QRC20","internal_id":""}
```

hmm I want to transfer `amount 10` - is it expected that `,"total_amount":"5","spent_by_me":"5"` ?

sergeyboyko0791 commented:
I think no.
The `total_amount` and `spent_by_me` fields now contain Qtum amount that will be sent. In other words, these are equal sum of Qtum outputs. Should the fields contain QRC20 spent amount?

artemii235 commented:
@sergeyboyko0791 
> Should the fields contain QRC20 spent amount?

Yeah, the intention of these fields is displaying amounts of the *actual* coin being transferred. The amount of spend QTUM is displayed in `fee_details`.
I've also noticed the small change to be made in regards to `fee_details`, it displays the name of the token, not QTUM:
`"fee_details":{"coin":"QRC20","miner_fee":"0.10000299","gas_limit":250000,"gas_price":40,"total_gas_fee":"0.1"}`
Should be
`"fee_details":{"coin":"QTUM","miner_fee":"0.10000299","gas_limit":250000,"gas_price":40,"total_gas_fee":"0.1"}`
because all fees are paid in QTUM.

sergeyboyko0791 commented:
I've changed the fields within withdraw response, they contain values in QRC20.
Also I've replaced QRC20 (ticker) with QTUM (platform) within FeeDetails.

tonymorony commented:
now output is clear, thanks!

```json
curl --url "http://127.0.0.1:7783" --data "{\"method\":\"withdraw\",\"coin\":\"QRC20\",\"to\":\"qMHa3L2VG9ZHJYQSZmCpv5PCvtxXfFEyqJ\",\"amount\":\"10\",\"userpass\":\"$userpass\"}"
{
  "tx_hex": "010000000276cf454685553e1ddce276e4d86432c17599a47b03fa8fc414525e1aac3bacdc020000006a47304402204dfdecacaba6bbb720f2bcd467196475843811a98fccf97193f4821fd932eed40220303e9a85fccbf9ee52807dde6b05e48e2b638ac952acb1493ad071468381a83f012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff9688d30f7e9cab2028e3802e209da8ca1047971e5881bd37f159ea2ff53e5c31010000006b483045022100c60746b494610fad9731609f519d6722ac3990d8048d581c1388cae3e7922835022050c7f430d7c1156aca4059516973fc09e987e4321cb22c60582c2ce9cc6797e6012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff02000000000000000062540390d003012844a9059cbb00000000000000000000000028e2ef01b3046a4eaaae78640285a29089bc0b46000000000000000000000000000000000000000000000000000000003b9aca00146029bddd12a5a828c89b61dc8c1d607d7e52b182c2be7c151d000000001976a914a19f7aa254dced0a17920ab1c2acf698841f67f688ac2638fd5e",
  "tx_hash": "2d73c957ac263ed19e7d2869fc86a7db6631e6e2cdbbd01d49af801820f4560e",
  "from": [
    "qYHxziMubYKZVwbKwrsn3YUhW4iL927NAS"
  ],
  "to": [
    "qMHa3L2VG9ZHJYQSZmCpv5PCvtxXfFEyqJ"
  ],
  "total_amount": "10",
  "spent_by_me": "10",
  "received_by_me": "0",
  "my_balance_change": "-10",
  "block_height": 0,
  "timestamp": 1593653286,
  "fee_details": {
    "coin": "QTUM",
    "miner_fee": "0.10000447",
    "gas_limit": 250000,
    "gas_price": 40,
    "total_gas_fee": "0.1"
  },
  "coin": "QRC20",
  "internal_id": ""
}
```

faced tx accepting problem on broadcasting @artemii235 (`Object({\"code\": Number(1), \"message\": String(\"the transaction was rejected by network rules.\\n\\nbad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 418 (code 16)`) :

```json
curl --url "http://127.0.0.1:7783" --data "{\"method\":\"send_raw_transaction\",\"coin\":\"QRC20\",\"tx_hex\":\"010000000276cf454685553e1ddce276e4d86432c17599a47b03fa8fc414525e1aac3bacdc020000006a47304402204dfdecacaba6bbb720f2bcd467196475843811a98fccf97193f4821fd932eed40220303e9a85fccbf9ee52807dde6b05e48e2b638ac952acb1493ad071468381a83f012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff9688d30f7e9cab2028e3802e209da8ca1047971e5881bd37f159ea2ff53e5c31010000006b483045022100c60746b494610fad9731609f519d6722ac3990d8048d581c1388cae3e7922835022050c7f430d7c1156aca4059516973fc09e987e4321cb22c60582c2ce9cc6797e6012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff02000000000000000062540390d003012844a9059cbb00000000000000000000000028e2ef01b3046a4eaaae78640285a29089bc0b46000000000000000000000000000000000000000000000000000000003b9aca00146029bddd12a5a828c89b61dc8c1d607d7e52b182c2be7c151d000000001976a914a19f7aa254dced0a17920ab1c2acf698841f67f688ac2638fd5e\",\"userpass\":\"$userpass\"}"

{
  "error": "rpc:305] lp_coins:737] utxo_common:924] JsonRpcError { client_info: \"coin: QRC20\", request: JsonRpcRequest { jsonrpc: \"2.0\", id: \"23\", method: \"blockchain.transaction.broadcast\", params: [String(\"010000000276cf454685553e1ddce276e4d86432c17599a47b03fa8fc414525e1aac3bacdc020000006a47304402204dfdecacaba6bbb720f2bcd467196475843811a98fccf97193f4821fd932eed40220303e9a85fccbf9ee52807dde6b05e48e2b638ac952acb1493ad071468381a83f012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff9688d30f7e9cab2028e3802e209da8ca1047971e5881bd37f159ea2ff53e5c31010000006b483045022100c60746b494610fad9731609f519d6722ac3990d8048d581c1388cae3e7922835022050c7f430d7c1156aca4059516973fc09e987e4321cb22c60582c2ce9cc6797e6012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff02000000000000000062540390d003012844a9059cbb00000000000000000000000028e2ef01b3046a4eaaae78640285a29089bc0b46000000000000000000000000000000000000000000000000000000003b9aca00146029bddd12a5a828c89b61dc8c1d607d7e52b182c2be7c151d000000001976a914a19f7aa254dced0a17920ab1c2acf698841f67f688ac2638fd5e\")] }, error: Response(95.217.83.126:10001, Object({\"code\": Number(1), \"message\": String(\"the transaction was rejected by network rules.\\n\\nbad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 418 (code 16)\\n[010000000276cf454685553e1ddce276e4d86432c17599a47b03fa8fc414525e1aac3bacdc020000006a47304402204dfdecacaba6bbb720f2bcd467196475843811a98fccf97193f4821fd932eed40220303e9a85fccbf9ee52807dde6b05e48e2b638ac952acb1493ad071468381a83f012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff9688d30f7e9cab2028e3802e209da8ca1047971e5881bd37f159ea2ff53e5c31010000006b483045022100c60746b494610fad9731609f519d6722ac3990d8048d581c1388cae3e7922835022050c7f430d7c1156aca4059516973fc09e987e4321cb22c60582c2ce9cc6797e6012103aa464529c190398d44017c3d7a71a854ad195a9d396bf84182679c242bb60085ffffffff02000000000000000062540390d003012844a9059cbb00000000000000000000000028e2ef01b3046a4eaaae78640285a29089bc0b46000000000000000000000000000000000000000000000000000000003b9aca00146029bddd12a5a828c89b61dc8c1d607d7e52b182c2be7c151d000000001976a914a19f7aa254dced0a17920ab1c2acf698841f67f688ac2638fd5e]\")})) }"
}
```

sergeyboyko0791 commented:
The mature confirmations limit is 500 for Qtum. By default this limit is 100 similar to Bitcoin. I've forgotten to note about this. You can specify the limit within config:

`{ "coin": "QRC20", "mature_confirmations": 500 }`

tonymorony commented:
Thanks! After `"required_confirmations": 0,` in settings I was able to send one tx after another. However, to send multiple token txs in single block manual split for QTUM utxos is needed because when my first token tx in mempool I'm getting such error: 

```
curl --url "http://127.0.0.1:7783" --data "{\"method\":\"withdraw\",\"coin\":\"QRC20\",\"to\":\"qMHa3L2VG9ZHJYQSZmCpv5PCvtxXfFEyqJ\",\"amount\":\"10\",\"userpass\":\"$userpass\"}"
{"error":"rpc:305] lp_coins:724] qrc20:490] utxo_common:236] Couldn't generate tx from empty utxos set"}
```

I have one concern: noticed that changes in that PR are "hardforking" for https://raw.githubusercontent.com/KomodoPlatform/coins/master/coins file usage for all coins activation 

```
curl --url "http://127.0.0.1:7783" --data "{\"userpass\":\"$userpass\",\"method\":\"electrum\",\"coin\":\"BCH\",\"servers\":[{\"url\":\"bch.imaginary.cash:50001\"}],\"mm2\":1}"
{"error":"rpc:305] lp_commands:74] lp_coins:661] invalid type: null, expected string or map"}
```

maybe we can change error message in such case on more informative like a "coins file you use is deprecated, please execute ./mm2 update_config command to convert or download new one" ? 

Also, we should not forget to open PR into coins repo, update all GUIs and notify devs/traders community (just a reminder for us, I know that you keep it in mind :) )

sergeyboyko0791 commented:
> maybe we can change error message in such case on more informative like a "coins file you use is deprecated, please execute ./mm2 update_config command to convert or download new one" ?

Would the following error message satisfy?

`{"error":"rpc:305] lp_commands:74] lp_coins:661] Expect "protocol" field. Probably coins file you use is deprecated, please execute ./mm2 update_config command to convert or download new one"}`

It will be little difficult to avoid the prefix `rpc:305] lp_commands:74] lp_coins:661]`

artemii235 commented:
@sergeyboyko0791 I would reword it a bit: `"protocol" field is missing in coins file. The file format is deprecated, please execute ./mm2 update_config command to convert it or download a new one`.

artemii235 commented:
@gcharang Hi! Could you please add an "Upgrade guide" to AtomicDEX-API docs which will contain the steps needed to convert config file after MM2 update? The config converter example can be seen in `./mm2 help`. Please let me know if you need any assistance with it.

gcharang commented:
@artemii235 I added a doc explaining how to update the coins file. Please review and suggest any corrections/additions. https://github.com/KomodoPlatform/developer-docs/pull/193
